pallets/domains/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+    pub enum Event<T: Config> {
+        /// The domain name was successfully registered and stored.
+        DomainRegistered(<T as frame_system::pallet::Config>::AccountId, DomainName<T>/*, BalanceOf<T>*/),
+        /// The domain meta was successfully updated.
+        DomainUpdated(<T as frame_system::pallet::Config>::AccountId, DomainName<T>),
+        /// The domains list was successfully added to the reserved list.
+        DomainsReserved(u16),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The content stored in domain metadata was not changed.
+        DomainContentNotChanged,
+        /// Cannot insert that many domains to a storage at once.
+        DomainsInsertLimitReached,
+        /// Cannot register more than `MaxDomainsPerAccount` domains.
+        TooManyDomainsPerAccount,
+        /// The domain label may contain only A-Z, 0-9 and hyphen characters.
+        DomainContainsInvalidChar,
+        /// The domain label length must be between 7 and 63 characters, inclusive.
+        DomainNameIsTooShort,
+        /// The domain has expired.
+        DomainHasExpired,
+        /// Domain was not found by either custom domain name or top level domain.
+        DomainNotFound,
+        /// This domain cannot be registered yet, because it is reserved.
+        DomainIsReserved,
+        /// This domain is already held by another account.
+        DomainAlreadyOwned,
+        /// A new inner value is the same as the old one.
+        InnerValueNotChanged,
+        /// Lower than Second level domains are not allowed.
+        LowerLevelDomainsNotAllowed,
+        /// This account is not allowed to update the domain metadata.
+        NotDomainOwner,
+        /// A new outer value is the same as the old one.
+        OuterValueNotChanged,
+        /// Reservation period cannot be a zero value.
+        ZeroReservationPeriod,
+        /// Cannot store a domain for that long period of time.
+        TooBigRegistrationPeriod,
+        /// Top level domain must be specified.
+        TopLevelDomainNotSpecified,
+        /// Top level domain not allowed.
+        TopLevelDomainNotAllowed,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Registers a domain ([full_domain]) using origin with [content],
+        /// and set the domain to expire in [expires_in].
+        #[pallet::weight(<T as Config>::WeightInfo::register_domain())]
+        pub fn register_domain(
+            origin: OriginFor<T>,
+            full_domain: DomainName<T>,
+            content: Content,
+            expires_in: <T as frame_system::pallet::Config>::BlockNumber,
+        ) -> DispatchResult {
+            let owner = ensure_signed(origin)?;
+
+            Self::do_register_domain(owner, full_domain, content, expires_in, ReserveDeposit::Yes)?;
+
+            Ok(())
+        }
+
+        /// Registers a domain ([full_domain]) using root in behalf of a [target] with [content],
```

Comment:
Copy all suggestions to docs of the previous extrinsic.

```suggestion
        /// Registers a domain ([full_domain]) using root on behalf of a [target] with [content],
```